internal/discovery/discovery.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog/v2"
+)
+
+// Interval is the interval at which the discovery client fetches the list of objects.
+const Interval = 30 * time.Second
+
+// Discover starts the discovery process, fetching all the objects that can be listed from the apiserver, every `DiscoveryInterval` seconds.
+// ResolveGVK needs to be called after Discover to generate factories.
+func (r *GVKMap) Discover(discoveryClient *discovery.DiscoveryClient) {
+	groupList, err := discoveryClient.ServerGroups()
+	if err != nil {
+		klog.Errorln("Failed to fetch server groups:", err)
+	}
+	// Reset the map difference.
+	r.MapDifference = []GroupVersionKind{}
+	gvkMap := make(map[string]map[string][]string)
+	for _, group := range groupList.Groups {
+		gvkMap[group.Name] = make(map[string][]string)
+		for _, version := range group.Versions {
+			gvkMap[group.Name][version.Version] = make([]string, 0)
+			resources, err := discoveryClient.ServerResourcesForGroupVersion(version.GroupVersion)
+			if err != nil {
+				klog.Errorln("Failed to fetch server resources for group version:", err)
+			}
+			for _, resource := range resources.APIResources {
+				gvkMap[group.Name][version.Version] = append(gvkMap[group.Name][version.Version], resource.Kind)
+				if r.OldMap != nil {
+					arrayHasElement := func(array []string, element string) bool {
+						for _, v := range array {
+							if v == element {
+								return true
+							}
+						}
+						return false
+					}
```

Comment:
I would move this function declaration out of the for loop, so that you only need to instantiate the function a single time.